### PR TITLE
Feature/academic work reclassify changes

### PR DESCRIPTION
**Pull request by RishiDiwanTT**
## Description
Switched around the query to focus on subject id
    
Rather than the unindexed subject.checked which would cause a table scans
across all joined tables
This is FAR more efficient than the previous implementation, although it is not as
"succinct" since we are doing 2 different queries
First we yield subjects with checked = false one a time, then we query for works
with only that subject.id
Even if we were to index subject.checked, it would make little difference as
a boolean column only has 2 possible values

## Motivation and Context
The primary looping query for the script `ReclassifyWorksForUncheckedSubjectsScript` was causing performance issues
on the Theta deployment.
This is an exercise in optimisation

## How Has This Been Tested?
Manually ran the script, and wrote new unit tests

## Checklist

- [ ] I have updated the documentation accordingly.
- [x] All new and existing tests passed.

### CODE REVIEW

File: core/scripts.py [hunk 1]
```diff
@@ -1657,9 +1675,118 @@ class ReclassifyWorksForUncheckedSubjectsScript(WorkClassificationScript):
...
+            self._db.query(Subject)
+            .filter(Subject.checked == False)
+            .order_by(Subject.id)
+        )
+        last_id = None
+        while True:
+            qu = query
+            if last_id:
+                qu = qu.filter(Subject.id > last_id)
+            subject = qu.first()
```
[the remark this answers is not in the document]
Comment (RishiDiwanTT) [reply core/scripts.py:1725]: First() does this inherently
https://docs.sqlalchemy.org/en/13/orm/query.html#sqlalchemy.orm.query.Query.first

File: core/model/work.py [hunk 2]
```diff
@@ -1011,7 +1011,7 @@ def calculate_presentation(
             or classification_changed
             or summary != self.summary
             or summary_text != new_summary_text
-            or float(quality) != float(self.quality)
+            or float(quality or 0) != float(self.quality or 0)
```
[the remark this answers is not in the document]
Comment (RishiDiwanTT) [reply core/model/work.py:1014]: It is not guaranteed that default quality is available here, as its first instantiation is within an if-block

File: core/scripts.py [hunk 3]
```diff
@@ -1657,9 +1675,118 @@ class ReclassifyWorksForUncheckedSubjectsScript(WorkClassificationScript):
...
+            yield subject
+
+    def paginate_query(self, query) -> Generator:
+        """Page this query using the row-wise comparison
+        technique unique to this job. We have already ensured
+        the ordering of the rows follws all the joined tables"""
+
+        for subject in self._unchecked_subjects():
+
+            last_work: Optional[Work] = None  # Last work object of the previous page
```
[the remark this answers is not in the document]
Comment (RishiDiwanTT) [reply core/scripts.py:1740]: mypy gets a little grumpy

```
core/scripts.py:1740: error: Incompatible types in assignment (expression has type "None", variable has type "Work")
```

File: core/model/work.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tdilauro) [reply core/model/work.py:1014]: I should clarify my comment, which now seems so obviously ambiguous. I meant, should we change line 1014 to:
```
    or float(quality or default_quality) != float(self.quality or default_quality)
```
from
```
    or float(quality or 0) != float(self.quality or 0)
```
?

Also -- and at least as importantly -- allowing thecondition from line 1014 above to participate in the calculation of `changed` without regard for the value of `policy.calculate_quality` looks to me like a logic error. It seems like it should be something like:
```
    or (policy.calculate_quality and float(quality or default_quality) != float(self.quality or default_quality))
```

[the remark this answers is not in the document]
Comment (RishiDiwanTT) [reply core/model/work.py:1014]: Logically this seems sound, since `quality` only changes in one place

**Review by tdilauro — APPROVED** [review]
Looks good to me! 🏁

**Pull request merged**